src/main/java/com/puppycrawl/tools/checkstyle/checks/DescendantTokenCheck.java

```diff
@@ -156,7 +156,7 @@ public class DescendantTokenCheck extends AbstractCheck {
      */
     private boolean sumTokenCounts;
     /** Specify set of tokens with limited occurrences as descendants. */
-    @XdocsPropertyType(PropertyType.TOKEN_ARRAY)
+    @XdocsPropertyType(PropertyType.STRING_ARRAY)
     private int[] limitedTokens = CommonUtil.EMPTY_INT_ARRAY;
```

Comment:
Tokens are integer values internally.
Externally they look like string values, but they are actually name of int constant from TokenTypes.java

In xml all looks like a string, but modules has strickt type so xml config loader does conversion.
But conceptually modules can be configured in runtime by strick types.

Reply:
Please share problems that you have to let see how vwe can handle this problem

Reply:
While figuring out the `DescendantTokenCheck.java`, I saw the return type of property name limitedToken to be String array. However, I guess you have clarified that. 

```
 /**
     * Setter to specify set of tokens with limited occurrences as descendants.
     *
     * @param limitedTokensParam tokens to ignore.
     * @since 3.2
     */
    public void setLimitedTokens(String... limitedTokensParam) {
    ....
    }
```

Reply:
The question remains, if i go without the recent changes, the xml representation looks like below:
`<td><a href="../../property_types.html#subset of tokens TokenTypes">subset of tokens TokenTypes</a></td`

and, it is extracting `TOKEN_ARRAY` from `PropertyType.java`. 

The issue remains that for it to match with before macros migration i.e. 
```
            <tr>
              <td>limitedTokens</td>
              <td>Specify set of tokens with limited occurrences as descendants.</td>
              <td>
                subset of tokens
                <a href="../../apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html">TokenTypes</a>
              </td>
              <td><code>{}</code></td>
              <td>3.2</td>
            </tr>
```
But, to match it to this, the code doesn't go inside the below section of `writePropertyTypeCell` method. 
```
            else {
                final List<String> configurableTokens = SiteUtil
                        .getDifference(check.getAcceptableTokens(),
                                check.getRequiredTokens())
                        .stream()
                        .map(TokenUtil::getTokenName)
                        .collect(Collectors.toList());
                sink.text("subset of tokens");

                writeTokensList(sink, configurableTokens, SiteUtil.PATH_TO_TOKEN_TYPES);
            }
```
Because it is failing this condition: `if (SiteUtil.TOKENS.equals(propertyName))`

One approach would be to have a specific use-case of limitedTokens, and create a condition to match it with the original xdocs. I hope I didn't confuse you. Please let me know if anything wasn't clear from my side.

Reply:
What is the relevance of the below code? Any idea?
```
private static final Set<String> NON_BASE_TOKEN_PROPERTIES = Collections.unmodifiableSet(
      Arrays.stream(new String[] {
          "AtclauseOrder - target",
          "DescendantToken - limitedTokens",
          "IllegalType - memberModifiers",
          "MagicNumber - constantWaiverParentToken",
          "MultipleStringLiterals - ignoreOccurrenceContext",
      }).collect(Collectors.toSet()));
```

Reply:
> Hmm, this actually means it should be String[].
> 
> https://github.com/checkstyle/checkstyle/blob/0f9d09e1d27c01d2fbdefd6c035dc4f4a9a97cb7/src/main/java/com/puppycrawl/tools/checkstyle/checks/DescendantTokenCheck.java#L304
> 
> All others have it as in[] . So converter is required. But here look like we missed it, and it should be String[]. As type of properties is API, we can not break it.

About the converter, can you guide me further? And, what is the way forward from here?

Reply:
My learnings so far: 
```
    @XdocsPropertyType(PropertyType.STRING_ARRAY)
    private int[] limitedTokens = CommonUtil.EMPTY_INT_ARRAY;
```
it retuns int[], which possibly is the **inconsistency** highlighted. And, changing it to `String[]` might lead to **issues in backward compatibility** by breaking things. 

About, the **conversion**, i assume it is already being done in the setter method of `setLimitedTokens`. 
```
    public void setLimitedTokens(String[] limitedTokensParam) {
        limitedTokens = new int[limitedTokensParam.length];

        int maxToken = 0;
        for (int i = 0; i < limitedTokensParam.length; i++) {
            limitedTokens[i] = TokenUtil.getTokenId(limitedTokensParam[i]);
            if (limitedTokens[i] >= maxToken + 1) {
                maxToken = limitedTokens[i];
            }
        }
        counts = new int[maxToken];
    }
```
When i inspect the custom annotation of `XdocsPropertyType`, the javadoc comments highlight something
```
/**
 * The type of property used in documentation and configuration files
 * for clarification of the user-friendly type.
 * For example, some string array property may represent a set of tokens.
 */

```
So, i assume the statment `some string array property may represent a set of tokens` hinted towards the original code i.e. `@XdocsPropertyType(PropertyType.TOKENS_ARRAY)`. 

Based on maintainers further guidance, i will proceed with what the ideal xml be like. And, please correct me if i was wrong somewhere in the above learnings.

Reply:
@Romani, do you want me to replace the var-args for this?
Original:
`public void setLimitedTokens(String... limitedTokensParam) `
Possible requested change:
`public void setLimitedTokens(String[] limitedTokensParam) `

Reply:
Metadata 
https://github.com/checkstyle/checkstyle/blob/b31faf461e8fe543db5791462888ed26de017c93/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/DescendantTokenCheck.xml#L19-23

Already stated it is array of string type, so our direction to `String[]` is good.

Change of `String...` to `String[]` is not going to do difference. It is same.
I think nothing is required to be changed.